react/week-21/lesson.md

```diff
@@ -327,100 +327,114 @@ render() {
-
-In contrast, we can get more control over our input data by using the *controlled component* pattern. Let's look at an example ([interactive example](https://codesandbox.io/s/4jq1yqy8kx)):
-
-```js
-class ControlledComponent extends Component {
+class CreateAccountForm extends Component {
   constructor(props) {
     super(props);
     this.state = {
-      name: ''
+      username: "",
+      email: "",
+      password: ""
     };
   }
 
-  handleChange = (event) => {
+  handleChange = event => {
     this.setState({
-      name: event.target.value
+      [event.target.name]: event.target.value
     });
   };
 
+  submit = () => {
+    console.log("Do something with the form values...");
+    console.log(`Username = ${this.state.username}`);
+    console.log(`Email = ${this.state.email}`);
+    console.log(`Password = ${this.state.password}`);
+  };
+
   render() {
     return (
-      <input
-        type="text"
-        placeholder="Name"
-        value={this.state.name}
-        onChange={this.handleChange}
-      />
+      <div>
+        <div>
+          <input
+            type="text"
+            name="username"
+            placeholder="Username"
+            value={this.state.username}
+            onChange={this.handleChange}
+          />
+        </div>
+        <div>
+          <input
+            type="text"
+            name="email"
+            placeholder="Email"
+            value={this.state.email}
+            onChange={this.handleChange}
+          />
+        </div>
+        <div>
+          <input
+            type="password"
+            name="password"
+            placeholder="Password"
+            value={this.state.password}
+            onChange={this.handleChange}
+          />
+        </div>
+        <button type="submit" onClick={this.submit}>
+          Create account
+        </button>
+      </div>
     );
   }
 }
 ```
 
-Now we're controlling the `value` of the input by using the value from state. This means that we can only change the value by updating state. If you didn't call `this.setState()` in the `handleChange` method, then the input's value would never change and it would appear as if you couldn't type in the input!
-
-This pattern is useful if you need to keep track of what the user is typing in the input. We could transform the string before we set it with `this.setState()`, for example by calling `toUpperCase()` on the string.
+We now have three different inputs named `username`, `email` and `password`, and we keep each entered value in a state with the same name. The method `handleChange` is reused to keep track of each change of value. The trick here is to use the name of the input element to update the corresponding state. Finally, when the user clicks on the submit button, the `submit` method is called to process the values. They are currently just displayed in the console but you could imagine validating the format of these values and sending them in a POST request.
```

Comment:
Previously I considered the refs/forms section "bonus material" that would only be covered if we had time in class - that's why it's at the end here.

But now I feel like it's slightly more important, so might be worth adding a few exercises?

Reply:
I think the form part with controlled component is important and really need to be seen in class, students are going to use it a lot for their final group projects (at least they did in Glasgow). But they never really used Refs for example and we never really pushed them in that direction either. I can add more exercises following this.